service/history/replication/ack_manager.go

```diff
@@ -445,351 +441,38 @@ func (p *ackMgrImpl) generateSyncActivityTask(
-				TaskType:     enumsspb.REPLICATION_TASK_TYPE_SYNC_ACTIVITY_TASK,
-				SourceTaskId: taskID,
-				Attributes: &replicationspb.ReplicationTask_SyncActivityTaskAttributes{
-					SyncActivityTaskAttributes: &replicationspb.SyncActivityTaskAttributes{
-						NamespaceId:        namespaceID.String(),
-						WorkflowId:         workflowID,
-						RunId:              runID,
-						Version:            activityInfo.Version,
-						ScheduledEventId:   activityInfo.ScheduledEventId,
-						ScheduledTime:      activityInfo.ScheduledTime,
-						StartedEventId:     activityInfo.StartedEventId,
-						StartedTime:        startedTime,
-						LastHeartbeatTime:  activityInfo.LastHeartbeatUpdateTime,
-						Details:            activityInfo.LastHeartbeatDetails,
-						Attempt:            activityInfo.Attempt,
-						LastFailure:        activityInfo.RetryLastFailure,
-						LastWorkerIdentity: activityInfo.RetryLastWorkerIdentity,
-						VersionHistory:     versionhistory.CopyVersionHistory(currentVersionHistory),
-					},
-				},
-				VisibilityTime: &taskInfo.VisibilityTimestamp,
-			}, nil
-		},
+		taskInfo,
+		p.workflowCache,
 	)
 }
 
 func (p *ackMgrImpl) generateHistoryReplicationTask(
 	ctx context.Context,
 	taskInfo *tasks.HistoryReplicationTask,
 ) (*replicationspb.ReplicationTask, error) {
-	namespaceID := namespace.ID(taskInfo.NamespaceID)
-	workflowID := taskInfo.WorkflowID
-	runID := taskInfo.RunID
-	taskID := taskInfo.TaskID
-	replicationTask, err := p.processReplication(
-		ctx,
-		true, // still necessary to send out history replication message if workflow closed
-		namespaceID,
-		workflowID,
-		runID,
-		func(mutableState workflow.MutableState) (*replicationspb.ReplicationTask, error) {
-			versionHistoryItems, branchToken, err := getVersionHistoryItems(
-				mutableState,
-				taskInfo.FirstEventID,
-				taskInfo.Version,
-			)
-			if err != nil {
-				return nil, err
-			}
-
-			eventsBlob, err := p.getEventsBlob(
-				ctx,
-				branchToken,
-				taskInfo.FirstEventID,
-				taskInfo.NextEventID,
-			)
-			if err != nil {
-				return nil, p.handleReadHistoryError(namespaceID, workflowID, runID, err)
-			}
-
-			replicationTask := &replicationspb.ReplicationTask{
-				TaskType:     enumsspb.REPLICATION_TASK_TYPE_HISTORY_V2_TASK,
-				SourceTaskId: taskID,
-				Attributes: &replicationspb.ReplicationTask_HistoryTaskAttributes{
-					HistoryTaskAttributes: &replicationspb.HistoryTaskAttributes{
-						NamespaceId:         namespaceID.String(),
-						WorkflowId:          workflowID,
-						RunId:               runID,
-						VersionHistoryItems: versionHistoryItems,
-						Events:              eventsBlob,
-						// NewRunEvents will be set in processNewRunReplication
-					},
-				},
-				VisibilityTime: &taskInfo.VisibilityTimestamp,
-			}
-			return replicationTask, nil
-		},
-	)
-	if err != nil {
-		return replicationTask, err
-	}
-	return p.processNewRunReplication(
+	return convertHistoryReplicationTask(
```

Comment:
nit: maybe call the underlying method directly?

Reply:
long term: the "ack manger" should be removed, so not worth making more changes